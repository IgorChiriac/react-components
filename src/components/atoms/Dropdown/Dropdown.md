Styled Dropdown component.

It consists of 2 components:

- `<Dropdown />`: which extends from html `<select>` element
- `<Option />`: which extends from html `<option>` element

```jsx
import { Option } from './Dropdown';

<Dropdown name="I ❤️ dropdowns" defaultValue="third">
  <Option value="first">First value</Option>
  <Option value="second">Second value</Option>
  <Option value="third">Third value</Option>
  <Option value="fourth">This value is really quite long</Option>
</Dropdown>;
```

#### hasError

```jsx
import { Option } from './Dropdown';

<Dropdown hasError={true}>
  <Option value="first">First value</Option>
</Dropdown>;
```
