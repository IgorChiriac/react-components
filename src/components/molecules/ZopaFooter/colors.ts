export const neutral900 = '#1C2139';
export const neutral200 = '#A9ACBA';
export const neutral400 = '#6F748B';
export const neutral800 = '#282F52';
