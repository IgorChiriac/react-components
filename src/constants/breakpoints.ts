export const breakpoints = {
  desktop: 1280,
  phone: 600,
  tablet: 720,
} as const;

export type TDeviceSizes = keyof typeof breakpoints;
