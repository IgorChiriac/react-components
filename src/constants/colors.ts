export type TTextHexColors = '#FFFFFF' | '#C5C6CD' | '#0D0A38' | '#1EC06A' | '#EE0505';
export type THeadingHexColors = '#FFFFFF' | '#0D0A38';
export type TLinkHexColors = '#FFFFFF' | '#0D0A38';

export const colors = {
  base: {
    primary: '#00B9A7',
    secondary: '#4A3EDE',
  },
  neutral: {
    white: '#FFFFFF',
    light: '#EAEAEE',
    medium: '#C5C6CD',
    dark: '#0D0A38',
  },
  semantic: {
    success: '#1EC06A',
    alert: '#FFB428',
    error: '#EE0505',
  },
};
