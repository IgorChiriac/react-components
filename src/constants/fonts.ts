export const openSans: string =
  '"Open Sans", -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol"';
export const alverata: string = 'Alverata, "Arial Black", Verdana, sans-serif';
