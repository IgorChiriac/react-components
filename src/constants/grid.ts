const grid = {
  breakpoints: {
    l: 992,
    m: 768,
    s: 576,
    xl: 1200,
    xs: 0,
  },
  cols: 12,
  gutter: 16,
  width: {
    l: 960,
    m: 720,
    s: 540,
    xl: 1140,
  },
};

export default grid;
