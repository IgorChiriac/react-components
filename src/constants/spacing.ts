export const spacing = [0, 4, 8, 16, 24, 32, 40, 56, 64, 104];
