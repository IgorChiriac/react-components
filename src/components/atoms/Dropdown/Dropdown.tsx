import React from 'react';
import styled from 'styled-components';
import * as colors from '../../../constants/colors';
import chevronDown from '../../../content/images/chevron-down.svg';
import { getBorderColorByStatus } from '../../../helpers/utils';

export const DEFAULT_COLOR = colors.neutral.neutral75;

export interface IDropdownProps extends React.SelectHTMLAttributes<HTMLSelectElement> {
  /**
   * Border gets red if this is set to true.
   */
  hasError?: boolean;
  /**
   * Border gets green if this is set to true.
   */
  isValid?: boolean;
}

export interface IOption extends React.OptionHTMLAttributes<HTMLOptionElement> {}

const Option: React.FunctionComponent<IOption> = styled.option``;
const Dropdown: React.FunctionComponent<IDropdownProps> = styled.select<IDropdownProps>`
  appearance: none;

  background: transparent url(${chevronDown}) no-repeat calc(100% - 13px) center;
  background-size: 13px;

  border: 2px solid ${getBorderColorByStatus};
  border-radius: 4px;
  padding: 8px 32px 8px 16px;
  font-size: 20px;
  height: 46px;
  min-width: 100px;

  &:focus {
    outline-width: 0;
    border: 2px solid ${colors.extended.blue500};
    transition: border 0.2s;
  }
`;

export default Dropdown;
export { Option };
