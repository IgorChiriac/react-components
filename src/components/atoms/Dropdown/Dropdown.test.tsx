import React from 'react';
import { render } from '@testing-library/react';
import { colors } from '../../../constants/colors';
import Dropdown, { DEFAULT_COLOR, Option } from './Dropdown';

describe('<Dropdown />', () => {
  it('renders the Dropdown with options', () => {
    const { container } = render(
      <Dropdown>
        <Option value="first">First value</Option>
        <Option value="second">Second value</Option>
      </Dropdown>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders the Dropdown with one option', () => {
    const { container } = render(
      <Dropdown>
        <Option value="first">First value</Option>
      </Dropdown>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders the Dropdown with options and a default selected value', () => {
    const { container } = render(
      <Dropdown defaultValue="defaultValue">
        <Option value="otherValue">Other Value</Option>
        <Option value="defaultValue">Default value to be selected</Option>
      </Dropdown>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders the Dropdown with options passed through an array.map', () => {
    const options = [{ value: 'value1', text: 'value1' }, { value: 'value2', text: 'value2' }];
    const { container } = render(
      <Dropdown defaultValue="defaultValue">
        <Option value="">Select a value</Option>
        {options.map(({ value, text }) => (
          <Option key={value} value={value}>
            {text}
          </Option>
        ))}
      </Dropdown>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test.each`
    hasError     | expectBorderColor
    ${false}     | ${DEFAULT_COLOR}
    ${undefined} | ${DEFAULT_COLOR}
    ${true}      | ${colors.semantic.error}
  `(
    'renders the border with the "$expectBorderColor" color if the hasError is "$hasError"',
    ({ hasError, expectBorderColor }) => {
      const { container } = render(
        <Dropdown hasError={hasError}>
          <Option value="first">First value</Option>
          <Option value="second">Second value</Option>
        </Dropdown>,
      );
      expect(container.firstChild).toHaveStyleRule('border', expect.stringContaining(expectBorderColor));
    },
  );
});
