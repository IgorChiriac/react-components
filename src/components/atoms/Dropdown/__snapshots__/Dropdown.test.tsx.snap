// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown /> renders the Dropdown with one option 1`] = `
.c0 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: #fff url(chevron-down.svg) no-repeat calc(100% - 13px) center;
  background-size: 13px;
  border: 2px solid #D6D7DE;
  border-radius: 4px;
  padding: 8px 32px 8px 16px;
  font-size: 20px;
  height: 46px;
  min-width: 100px;
}

.c0:focus {
  outline-width: 0;
  border: 2px solid #4B3CFA;
  -webkit-transition: border 0.2s;
  transition: border 0.2s;
}

<select
  class="c0"
>
  <option
    value="first"
  >
    First value
  </option>
</select>
`;

exports[`<Dropdown /> renders the Dropdown with options 1`] = `
.c0 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: #fff url(chevron-down.svg) no-repeat calc(100% - 13px) center;
  background-size: 13px;
  border: 2px solid #D6D7DE;
  border-radius: 4px;
  padding: 8px 32px 8px 16px;
  font-size: 20px;
  height: 46px;
  min-width: 100px;
}

.c0:focus {
  outline-width: 0;
  border: 2px solid #4B3CFA;
  -webkit-transition: border 0.2s;
  transition: border 0.2s;
}

<select
  class="c0"
>
  <option
    value="first"
  >
    First value
  </option>
  <option
    value="second"
  >
    Second value
  </option>
</select>
`;

exports[`<Dropdown /> renders the Dropdown with options and a default selected value 1`] = `
.c0 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: #fff url(chevron-down.svg) no-repeat calc(100% - 13px) center;
  background-size: 13px;
  border: 2px solid #D6D7DE;
  border-radius: 4px;
  padding: 8px 32px 8px 16px;
  font-size: 20px;
  height: 46px;
  min-width: 100px;
}

.c0:focus {
  outline-width: 0;
  border: 2px solid #4B3CFA;
  -webkit-transition: border 0.2s;
  transition: border 0.2s;
}

<select
  class="c0"
>
  <option
    value="otherValue"
  >
    Other Value
  </option>
  <option
    selected=""
    value="defaultValue"
  >
    Default value to be selected
  </option>
</select>
`;

exports[`<Dropdown /> renders the Dropdown with options passed through an array.map 1`] = `
.c0 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: #fff url(chevron-down.svg) no-repeat calc(100% - 13px) center;
  background-size: 13px;
  border: 2px solid #D6D7DE;
  border-radius: 4px;
  padding: 8px 32px 8px 16px;
  font-size: 20px;
  height: 46px;
  min-width: 100px;
}

.c0:focus {
  outline-width: 0;
  border: 2px solid #4B3CFA;
  -webkit-transition: border 0.2s;
  transition: border 0.2s;
}

<select
  class="c0"
>
  <option
    value=""
  >
    Select a value
  </option>
  <option
    value="value1"
  >
    value1
  </option>
  <option
    value="value2"
  >
    value2
  </option>
</select>
`;
